### ISSUE TITLE
isinstance wrongly narrows away Any

### ISSUE THREAD

**Issue by freundTech** [python/mypy#12290]
**Bug Report**

Mypy falsely narrows a type when using `isinstance` with a tuple containing a variable of type `Any`.
This only happens when there is at least one other element in the tuple.
Note that a Class might have type Any if it is imported from an untyped module.

**To Reproduce**

```python
from typing import Any

class A: 
    pass

B: Any  # (Or from untyped_lib import B)

m: Any

if isinstance(m, (A, B)):
    reveal_type(m)
else:
    reveal_type(m)
```

**Expected Behavior**

Mypy reports no error. `m` is of type `Any` in both branches

**Actual Behavior**

The second branch is reported as unreachable.

**Your Environment**

- Mypy version used: 0.931
- Mypy command-line flags: --warn-unreachable
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: 3.10

**Comment by flisboac** [comment]
After the `isinstance` guard, `m` has type `A | Any`.

I think that, in a sense, if a union type can be `Any` (which is the case here, even if there's a more specified/evident type as an option), the correct assumption would be that the guarded variable (`m`) is effectively just `Any`, because for all practical purposes you can never be sure if it has the more specific type. A better approach would be to make two separate `isinstance` calls instead:

```py
if isinstance(m, A):
  reveal_type(m)  # here, it's assuredly A
elif isinstance(m, B):
  reveal_type(m) # here, it's Any, because B's type is unspecified
else:
  reveal_type(m) # here, it's just Any
```

**Comment by KotlinIsland** [comment]
This is extremely sus and inconsistent:
```py
from typing import Any

B: Any

m: Any

if not isinstance(m, int | B):
    print(1)  # unreachable

if not (isinstance(m, int) or isinstance(m, B)):
    print(1)  # no error
```

**Comment by flisboac** [comment]
@KotlinIsland Initially, your snippet didn't make much sense to me, because at first I thought it should be `B = Any` (instead of `B: any`; considering that that's the syntax for type aliases). As I recall, for `B` to be useable in `isinstance`, it should be a type, a type alias, or a tuple of those. `isinstance` also accepts unions. As `B` is declared as simply a variable with type `Any`, my first guess was that this should not work, or that it was invalid.

And yet it does work, and is neither wrong nor illegal. But just because `B` is typed as `Any`, which can be a substitute for any type, including `type`, `Unions`s, etc. According to [PEP-484](https://peps.python.org/pep-0484/#the-any-type):

> A special kind of type is Any. Every type is consistent with Any. It can be considered a type that has all values and all methods.

If you change the example slightly, you get a more informative error:

```py
from typing import Any

B: str

m: Any

reveal_type(int | B)  # note: Revealed type is "types.UnionType"

if not isinstance(m, int | B):
    print(1)

# error: Argument 2 to "isinstance" has incompatible type "str"; expected "Union[type, UnionType, Tuple[Union[type, UnionType, Tuple[Any, ...]], ...]]"
if not (isinstance(m, int) or isinstance(m, B)):
    print(1)
```

So, I think there's no bug here, because by the book it's allowed to happen.

**Comment by KotlinIsland** [comment]
@flisboac Why did you edit my example to make `B` an instance of `str`? in this example `B` is an instance of `Any` but at runtime it is a `type[X]`.

**Comment by KotlinIsland** [comment]
This example is generating a very clear false positive, and is extremely inconsistent, with different behavior if the value is assigned to a variable or not.
```py
def get_type_to_check():
    return bool
def get_value_to_check():
    return 1

m = get_value_to_check()

if not isinstance(m, str | get_type_to_check()):
    print(1)  # error: Statement is unreachable

if not isinstance(get_value_to_check(), str | get_type_to_check()):
    print(1)  # no error
```
[playground](https://mypy-play.net/?mypy=latest&python=3.10&flags=warn-unreachable&gist=76521918f3afb58cc065dc80d6b0fb50)
